File: ios/HybridViewModel.swift
import NitroModules
import RiveRuntime

class HybridViewModel: HybridViewModelSpec {
  let viewModel: RiveDataBindingViewModel?

  init(viewModel: RiveDataBindingViewModel) {
    self.viewModel = viewModel
  }

  var propertyCount: Double { MainThread.run { Double(viewModel?.propertyCount ?? 0) } }

  var instanceCount: Double { MainThread.run { Double(viewModel?.instanceCount ?? 0) } }

  var modelName: String { MainThread.run { viewModel?.name ?? "" } }

  func createInstanceByIndex(index: Double) throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard index >= 0 else { return nil }
      guard let viewModel = viewModel,
            let vmi = viewModel.createInstance(fromIndex: UInt(index)) else { return nil }
      return HybridViewModelInstance(viewModelInstance: vmi)
    }
  }

  func createInstanceByName(name: String) throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard let viewModel = viewModel,
            let vmi = viewModel.createInstance(fromName: name) else { return nil }
      return HybridViewModelInstance(viewModelInstance: vmi)
    }
  }

  func createDefaultInstance() throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard let viewModel = viewModel,
            let vmi = viewModel.createDefaultInstance() else {
        return nil
      }
      return HybridViewModelInstance(viewModelInstance: vmi)
    }
  }

  func createInstance() throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard let viewModel = viewModel,
            let vmi = viewModel.createInstance() else { return nil }
      return HybridViewModelInstance(viewModelInstance: vmi)
    }
  }

  func getPropertyCountAsync() throws -> Promise<Double> {
    return Promise.onMain { self.propertyCount }
  }

  func getInstanceCountAsync() throws -> Promise<Double> {
    return Promise.onMain { self.instanceCount }
  }

  func createInstanceByNameAsync(name: String) throws -> Promise<(any HybridViewModelInstanceSpec)?> {
    return Promise.onMain { try self.createInstanceByName(name: name) }
  }

  func createDefaultInstanceAsync() throws -> Promise<(any HybridViewModelInstanceSpec)?> {
    return Promise.onMain { try self.createDefaultInstance() }
  }

  func createBlankInstanceAsync() throws -> Promise<(any HybridViewModelInstanceSpec)?> {
    return Promise.onMain { try self.createInstance() }
  }
}

File: android/src/legacy/java/com/margelo/nitro/rive/HybridViewModel.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import app.rive.runtime.kotlin.core.ViewModel
import app.rive.runtime.kotlin.core.errors.ViewModelException
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise

@Keep
@DoNotStrip
class HybridViewModel(private val viewModel: ViewModel) : HybridViewModelSpec() {
  override val propertyCount: Double
    get() = viewModel.propertyCount.toDouble()
  override val instanceCount: Double
    get() = viewModel.instanceCount.toDouble()
  override val modelName: String
    get() = viewModel.name

  override fun createInstanceByIndex(index: Double): HybridViewModelInstanceSpec? {
    if (index < 0) return null
    try {
      val vmi = viewModel.createInstanceFromIndex(index.toInt())
      return HybridViewModelInstance(vmi)
    } catch (e: ViewModelException) {
      return null
    }
  }

  override fun createInstanceByName(name: String): HybridViewModelInstanceSpec? {
    try {
      val vmi = viewModel.createInstanceFromName(name)
      return HybridViewModelInstance(vmi)
    } catch (e: ViewModelException) {
      return null
    }
  }

  override fun createDefaultInstance(): HybridViewModelInstanceSpec? {
    try {
      val vmi = viewModel.createDefaultInstance()
      return HybridViewModelInstance(vmi)
    } catch (e: ViewModelException) {
      return null
    }
  }

  override fun createInstance(): HybridViewModelInstanceSpec? {
    try {
      val vmi = viewModel.createBlankInstance()
      return HybridViewModelInstance(vmi)
    } catch (e: ViewModelException) {
      return null
    }
  }

  override fun getPropertiesAsync(): Promise<Array<ViewModelPropertyInfo>> {
    return Promise.rejected(UnsupportedOperationException("getPropertiesAsync is not supported on the legacy backend"))
  }

  // Main-hopped like HybridRiveFile's lookups — the legacy runtime is only
  // safe to touch on the main thread (see riveMainScope).
  override fun getPropertyCountAsync(): Promise<Double> {
    return Promise.async(riveMainScope) { propertyCount }
  }

  override fun getInstanceCountAsync(): Promise<Double> {
    return Promise.async(riveMainScope) { instanceCount }
  }

  override fun createInstanceByNameAsync(name: String): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async(riveMainScope) { createInstanceByName(name) }
  }

  override fun createDefaultInstanceAsync(): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async(riveMainScope) { createDefaultInstance() }
  }

  override fun createBlankInstanceAsync(): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async(riveMainScope) { createInstance() }
  }
}
